pkg/gds/replica.go

```diff
@@ -112,46 +118,209 @@ func (r *Replica) Shutdown() error {
+	seen := make(map[string]struct{})
+	fetch := make(map[string]struct{})
+
+	// Step 1a: Loop over all incoming objects
+incomingLoop:
+	for _, remoteObj := range in.Objects {
+		// Step 1b: Fetch data from database (determining if the data is not found)
+		// TODO: move object fetching logic back to Store; ensure tombstones are included.
+		var data []byte
+		if data, err = ldb.Get([]byte(remoteObj.Key), nil); err != nil {
+			if errors.Is(err, leveldb.ErrEntityNotFound) {
+				// This exists on the remote, but not locally; so add to fetch.
+				if !in.Partial {
+					fetch[remoteObj.Key] = struct{}{}
+				}
+				continue incomingLoop
+			} else {
+				// This is an unhandled error; log it and return replica requires repair
+				log.Error().Err(err).Msg("could not get key from leveldb")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+		}
+
+		// Step 1c: Get the localObj representation based on the object namespace
+		// TODO: create an interface that returns the object representation
+		var localObj *global.Object
+		switch remoteObj.Namespace {
+		case store.NamespaceVASPs:
+			vasp := &pb.VASP{}
+			if err = proto.Unmarshal(data, vasp); err != nil {
+				// This is an unhandled error; log it and return unavailable
+				log.Error().Err(err).Msg("could not unmarshal VASP from leveldb")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+
+			// VASP specific retrieval of object metadata
+			if localObj, _, err = models.GetMetadata(vasp); err != nil {
+				// This is an unhandled error; log it and return unavailable
+				log.Error().Err(err).Msg("could not get metadata from VASP")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+
+			// Ensure vasp is stored on the object in case it is sent back to remote
+			if localObj.Data, err = anypb.New(vasp); err != nil {
+				log.Error().Err(err).Msg("could not marshal VASP Any back to remote replica")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+
+		case store.NamespaceCertReqs:
+			certreq := &models.CertificateRequest{}
+			if err = proto.Unmarshal(data, certreq); err != nil {
+				// This is an unhandled error; log it and return unavailable
+				log.Error().Err(err).Msg("could not unmarshal CertificateRequest from leveldb")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+
+			// Certreq specific method to retrieve metadata
+			localObj = certreq.Metadata
+
+			// Ensure certreq is stored on the object in case it is sent back to remote
+			if localObj.Data, err = anypb.New(certreq); err != nil {
+				log.Error().Err(err).Msg("could not marshal CertificateRequest Any back to remote replica")
+				return nil, status.Error(codes.FailedPrecondition, "replica requires repair")
+			}
+		default:
+			// Log error but continue processing, foreign namespace will be ignored
```

Comment:
#TODO will we eventually want to figure out how to replicate new namespaces that the local doesn't know about? Maybe leave it to the honu upgrade?